src/components/TextField/TextArea.js

```diff
@@ -0,0 +1,279 @@
+import IconRequired from '../Icon/IconRequired';
+import { TextAreaAutoHeight } from './TextFieldEnums';
+import './TextArea.scss';
+
+const propTypes = {
+    className: PropTypes.string,
+    componentRef: PropTypes.func,
+    hasAutoHeight: ListPropType([
+        TextAreaAutoHeight.ENABLED,
+        TextAreaAutoHeight.DISABLED,
+    ]),
+    hasLabelAlways: PropTypes.bool,
+    isDisabled: PropTypes.bool,
+    isFullWidth: PropTypes.bool,
+    isReadOnly: PropTypes.bool,
+    isRequired: PropTypes.bool,
+    isValid: PropTypes.bool,
+    label: PropTypes.string,
+    name: PropTypes.string,
+    onBlur: PropTypes.func,
+    onChange: PropTypes.func,
+    onEnterKey: PropTypes.func,
+    onFocus: PropTypes.func,
+    onKeyDown: PropTypes.func,
+    onKeyPress: PropTypes.func,
+    onKeyUp: PropTypes.func,
+    placeholder: PropTypes.string,
+    rows: PropTypes.number,
+    style: StyleObjectPropType(),
+    tooltipError: PropTypes.oneOfType([
+        PropTypes.element,
+        PropTypes.string,
+    ]),
+    tooltipHint: PropTypes.oneOfType([
+        PropTypes.element,
+        PropTypes.string,
+    ]),
+    tooltipRequired: PropTypes.oneOfType([
+        PropTypes.element,
+        PropTypes.string,
+    ]),
+    value: PropTypes.string,
+};
+
+const defaultProps = {
+    className: null,
+    componentRef: null,
+    hasAutoHeight: TextAreaAutoHeight.DISABLED,
+    hasLabelAlways: false,
+    isDisabled: false,
+    isFullWidth: false,
+    isReadOnly: false,
+    isRequired: false,
+    isValid: null,
+    label: null,
+    name: null,
+    onBlur: null,
+    onChange: null,
+    onEnterKey: null,
+    onFocus: null,
+    onKeyDown: null,
+    onKeyPress: null,
+    onKeyUp: null,
+    placeholder: null,
+    rows: 1,
+    style: null,
+    tooltipError: null,
+    tooltipHint: null,
+    tooltipRequired: 'required',
+    value: '',
+};
+
+let lastInstanceId = 0;
+
+class TextArea extends Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            hasFocus: false,
+            hasMouseOver: false,
+            showTooltip: false,
+            value: props.value,
+        };
+
+        lastInstanceId += 1;
+        this.uid = `textarea-${lastInstanceId}`;
+    }
+
+    componentDidMount = () => {
+        autosize(this.inputRef);
+    }
+
+    handleMouseEnter = () => {
+        this.setState({ hasMouseOver: true });
+    }
+
+    handleMouseLeave = () => {
+        this.setState({ hasMouseOver: false });
+    }
+
+    handleInputRef = (ref) => {
+        const { componentRef } = this.props;
+        this.inputRef = ref;
+        if (componentRef) {
+            componentRef(ref);
+        }
+    }
+
+    handleInputBlur = (event) => {
+        const { onBlur } = this.props;
+        this.setState({
+            hasFocus: false,
+            showTooltip: false,
+        });
+        if (onBlur) {
+            onBlur(event);
+        }
```

Comment:
I'm not 100% sure about this, I see your point and it is safer, however, does it matter if the `onBlur` function is called in parallel with `setState`?

I can't see a scenario where that would be a problem... with that in mind I like the idea of keeping the components as fast as possible

@andyberry88 can I get your opinion on this?

Reply:
The problem is that they won't happen in parallel. It's very likely that `onBlur` will fire before state has been updated. Just search for "setState callback" and you'll find plenty of people having problems when not using callbacks.
This might be less of a problem with blur (although we can't imagine all the possible use cases) but is very likely with `onChange` - the `onChange` handler could fire before the internal `value` has been updated. This could lead to some weird behaviours. Even if current component code wouldn't have a problem with this discrepancy, managing state this way may cause issues in the future (there were bugs caused by this in m-dash-ui).

I think it's good practice to ensure internal state is updated before calling external functions.

Reply:
I agree with @jfrej on this. We should make sure the internal state has been set before triggering events etc. outside of the component.

Reply:
Ok chaps, I bow to your expertise... I'll get it changed :)

Reply:
https://media.giphy.com/media/SKjTHluaRvLG/giphy.gif